### PR TITLE
Address NaNs if SDPA is called with all values masked from query

### LINKED ISSUE #156707: MPS SDPA returns NaN when attention mask blocks all rows
### 🐛 Describe the bug

When using scaled_dot_product_attention on MPS with attention masks that block all attention for a position, the function returns NaNs instead of zeros. This creates inconsistent behavior between CPU and MPS backends.

## Root Cause
In `aten/src/ATen/native/mps/operations/Attention.mm`, the `sdpa_general_mps` function directly calls Metal's `softMaxWithTensor` without handling the edge case where all attention scores are `-inf`. This occurs when an attention mask has `False` for all positions that a query can attend to.

I think the issue is here, when passing all-`-inf` tensor to Metal softmax, which produces NaNs
https://github.com/pytorch/pytorch/blob/e2c9d8d6414927ce754bbc40b767edf103cf16da/aten/src/ATen/native/mps/operations/Attention.mm#L117

While CPU code explicitly checks for `tmp_max == -infinity` and returns zeros:
https://github.com/pytorch/pytorch/blob/e2c9d8d6414927ce754bbc40b767edf103cf16da/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L649-L653

## Test Case
```python
import torch
import torch.nn.functional as F

# Minimal reproduction
q = torch.randn(1, 1, 2, 4)
k = torch.randn(1, 1, 2, 4)
v = torch.randn(1, 1, 2, 4)
mask = torch.tensor([[[[False, False], [True, True]]]], dtype=torch.bool)

# CPU returns zeros
out_cpu = F.scaled_dot_product_attention(q, k, v, attn_mask=mask)
print("CPU:", out_cpu[0, 0, 0, :3].tolist())  # [0.0, 0.0, 0.0]

# MPS returns NaNs
out_mps = F.scaled_dot_product_attention(q.to('mps'), k.to('mps'), v.to('mps'), attn_mask=mask.to('mps'))
print("MPS:", out_mps[0, 0, 0, :3].tolist())  # [nan, nan, nan]
```

## Impact
This kind of situation happens when you have left-padded batches for generation in HF's transformers.

tagging the mps people: @jhavukainen @kulinseth @malfet

### Versions

PyTorch version: 2.7.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.5 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: version 4.0.2
Libc version: N/A

Python version: 3.13.5 | packaged by conda-forge | (main, Jun 16 2025, 08:24:05) [Clang 18.1.8 ] (64-bit runtime)
Python platform: macOS-15.5-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3

Versions of relevant libraries:
[pip3] numpy==2.3.1
[pip3] torch==2.7.1

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by jhavukainen**
Fixes #156707

Detect if all values along the softmax axis are infs and overwrite the outputs for those computations with zeros before the final matmul. The behavior should be aligned with the CPU implementation.

These types of cases where all values along the dimension in the attention mask are false leading to the undefined outputs in softmax occur with left padded batches for generation in HF transformers according to the original issue.

### CODE REVIEW

**Comment by drisspg** [comment]
@malfet  I dont really know enough about the mps implementation to tell if this is the correct way to do this, but this seems to treat any Inf as the same? This is how I updated the existing flash implementations: https://github.com/pytorch/pytorch/pull/131863

**Comment by jhavukainen** [comment]
I see what you mean. Yeah this version would end up catching both negative and positive infinities instead of just negative ones so I'll need to change that. I'll push an update today to fix that.

**Review by malfet — APPROVED** [review]
LGTM, but I wonder what are the performance implications of adding those checks?

**Comment by jhavukainen** [comment]
@malfet Yeah that's a fair question. I ran a couple of naive benchmarks to get an idea of the effect this would have. For smaller shapes it does cause a measurable difference but with more relevant sizes it disappears within the measurement variability quite quickly, so overall I'd estimate the check doesn't cost too much for the average use case. Here's the script I used with an M2 Max device:

```
import torch
import numpy as np
import time

L=2048
S=512
NH=128
HS=1024
dtype=torch.float32

q = torch.randn([1, NH, L, HS], dtype=dtype, device="mps")
k = torch.randn([1, NH, S, HS], dtype=q.dtype, device="mps")
v = torch.randn([1, NH, S, HS], dtype=q.dtype, device="mps")

op = torch.nn.functional.scaled_dot_product_attention

num_warmup = 5
n_stats = 50

for _ in range(num_warmup):
    out = op(q, k, v)
    torch.mps.synchronize()

stats = []
for i in range(n_stats):
    start = time.perf_counter_ns()
    out = op(q, k, v)
    torch.mps.synchronize()
    end = time.perf_counter_ns()
    stats.append((end-start)/1e3)

print(f"Took {np.mean(stats):0.2f}\u00B1{np.std(stats):0.2f} \u03bcs per iter")
```

And the results for two sets of shapes:
```
L=1, S=72, NH=32, HS=128
# Before change:
Took 310.55±37.74 μs per iter
# After change:
Took 368.98±35.13 μs per iter
```
```
L=2048, S=512, NH=128, HS=1024
# Before change:
Took 52839.62±147.31 μs per iter
# After change:
Took 52825.06±132.55 μs per iter
```


With the GPT2 train benchmark from torchbench I'm getting

```
Before change:
----------------------------------------------- benchmark 'hub': 1 tests -----------------------------------------------
Name (time in ms)                Min       Max      Mean  StdDev    Median     IQR  Outliers     OPS  Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------
test_train[hf_GPT2-mps]     398.4452  407.1690  400.6670  3.6971  398.8504  3.3065       1;1  2.4958       5           1
------------------------------------------------------------------------------------------------------------------------

After change:
----------------------------------------------- benchmark 'hub': 1 tests -----------------------------------------------
Name (time in ms)                Min       Max      Mean  StdDev    Median     IQR  Outliers     OPS  Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------
test_train[hf_GPT2-mps]     395.7955  400.3153  398.8589  1.7881  399.2772  1.8033       1;0  2.5072       5           1
------------------------------------------------------------------------------------------------------------------------
```
